prog/kubernetes/provision_kubernetes_node.rb

```diff
@@ -0,0 +1,134 @@
+    vm.sshable.cmd <<BASH
+  set -ueo pipefail
+  echo "net.ipv6.conf.all.forwarding=1\nnet.ipv6.conf.all.proxy_ndp=1\nnet.ipv4.conf.all.forwarding=1\nnet.ipv4.ip_forward=1" | sudo tee /etc/sysctl.d/72-clover-forward-packets.conf
+  sudo sysctl --system
+  sudo apt update
+  sudo apt install -y ca-certificates curl apt-transport-https gpg
+  sudo install -m 0755 -d /etc/apt/keyrings
+  sudo curl -fsSL https://download.docker.com/linux/ubuntu/gpg -o /etc/apt/keyrings/docker.asc
+  sudo chmod a+r /etc/apt/keyrings/docker.asc
+  echo "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/ubuntu \\$(. /etc/os-release && echo "\\$VERSION_CODENAME") stable" \\| sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
+  sudo apt update
+  sudo apt install -y containerd
+  sudo mkdir -p /etc/containerd
+  sudo touch /etc/containerd/config.toml
+  containerd config default | sudo tee /etc/containerd/config.toml
+  sudo sed -i 's/SystemdCgroup = false/SystemdCgroup = true/' /etc/containerd/config.toml
+  sudo systemctl restart containerd
+  curl -fsSL https://pkgs.k8s.io/core:/stable:/#{kubernetes_cluster.kubernetes_version}/deb/Release.key | sudo gpg --dearmor -o /etc/apt/keyrings/kubernetes-apt-keyring.gpg
+  echo 'deb [signed-by=/etc/apt/keyrings/kubernetes-apt-keyring.gpg] https://pkgs.k8s.io/core:/stable:/#{kubernetes_cluster.kubernetes_version}/deb/ /' | sudo tee /etc/apt/sources.list.d/kubernetes.list
+  sudo apt update
+  sudo apt install -y kubelet kubeadm kubectl
+  sudo apt-mark hold kubelet kubeadm kubectl
+  sudo systemctl enable --now kubelet
+BASH
+
+    hop_bootstrap_rhizome
+  end
+
+  label def bootstrap_rhizome
+    bud Prog::BootstrapRhizome, {"target_folder" => "kubernetes", "subject_id" => vm.id, "user" => "ubi"}
+
+    hop_wait_bootstrap_rhizome
+  end
+
+  label def wait_bootstrap_rhizome
+    reap
+    hop_assign_role if leaf?
+    donate
+  end
+
+  label def assign_role
+    write_hosts_file_if_needed
+
+    hop_init_cluster if kubernetes_cluster.cp_vms.count == 1
+
+    hop_join_control_plane
+  end
+
+  label def init_cluster
+    case vm.sshable.cmd("common/bin/daemonizer --check init_kubernetes_cluster")
+    when "Succeeded"
+      pop vm_id: vm.id
+    when "NotStarted"
+      ps = kubernetes_cluster.private_subnet
+      vm.sshable.cmd("common/bin/daemonizer '/home/ubi/kubernetes/bin/init-cluster #{kubernetes_cluster.name} #{kubernetes_cluster.endpoint} 443 #{ps.net4} #{ps.net6} #{vm.nics.first.private_ipv4}' init_kubernetes_cluster")
+      nap 30
+    when "InProgress"
+      nap 10
+    when "Failed"
+      Clog.emit("INIT CLUSTER FAILED")
+      nap 65536
+      # TODO: register deadline
+    end
+
+    nap 65536
+  end
+
+  label def join_control_plane
+    case vm.sshable.cmd("common/bin/daemonizer --check join_control_plane")
+    when "Succeeded"
+      pop vm_id: vm.id
+    when "NotStarted"
+      cp_sshable = kubernetes_cluster.cp_vms.first.sshable
```

Comment:
is it only required to write this new config into a single cp vm? What if this array returns the newly added VM, does it cause any issues?

Reply:
cp_vms are returned sorder by `created_at` so the `cp_vms.first` should already be a valid CP node. Running these commands against any CP node that's joined the cluster would be OK, but `cp_vms.first` is the safest one.

This design is not great when you think about concurrent operations. For future, instead of associating vms with clusters directly, we think about introducing a wrapper entity called KubernetesNode, that would act as a synchronization point as well.